File: src/main/java/ru/timakden/chesspuzzle/ChessPuzzle.java
package ru.timakden.chesspuzzle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Класс описывает решение шахматной задачи.
 */
class ChessPuzzle {
	/**
	 * Фабрика для создания шахматных фигур.
	 */
	private ChessPieceFactory chessPieceFactory = new ChessPieceFactory();

	/**
	 * Количество уникальных решений задачи с указанными параметрами.
	 */
	private int numberOfUniqueSolutions;

	/**
	 * Шахматная доска.
	 */
	private ChessBoard chessBoard;

	/**
	 * Ассоциативный массив для хранения последних размещённых шахматных фигур каждого типа.
	 */
	private Map<String, ChessPiece> lastPlacedChessPieces = new HashMap<>();

	ChessPuzzle(int rows, int columns, int kings, int queens, int rooks, int bishops, int knights) {
		numberOfUniqueSolutions = 0;

		List<String> remainingChessPieces = new ArrayList<>(kings + queens + bishops + rooks + knights);
		for (int i = 0; i < kings; i++) {
			remainingChessPieces.add("K");
		}
		for (int i = 0; i < queens; i++) {
			remainingChessPieces.add("Q");
		}
		for (int i = 0; i < rooks; i++) {
			remainingChessPieces.add("R");
		}
		for (int i = 0; i < bishops; i++) {
			remainingChessPieces.add("B");
		}
		for (int i = 0; i < knights; i++) {
			remainingChessPieces.add("N");
		}

		chessBoard = new ChessBoard(rows, columns, remainingChessPieces);
	}

	int getNumberOfUniqueSolutions() {
		return numberOfUniqueSolutions;
	}

	void solve() {
		List<String> remainingChessPieces = chessBoard.getRemainingChessPieces();

		// Берём первую доступную фигуру
		String chessPieceType = remainingChessPieces.remove(0);

		// Получаем последнюю поставленную фигуру такого типа
		ChessPiece lastPlacedChessPiece = lastPlacedChessPieces.get(chessPieceType);

		int r = 0;
		int c = 0;

		if (lastPlacedChessPiece != null) {
			// Пропускаем повторяющиеся решения
			r = lastPlacedChessPiece.getRow();
			c = lastPlacedChessPiece.getColumn() + 1;
		}

		for (int row = r; row < chessBoard.getRows(); row++) {
			for (int column = c; column < chessBoard.getColumns(); column++) {
				// Если клетка занята, то пропускаем итерацию
				if (!chessBoard.isCellFree(row, column)) {
					continue;
				}

				ChessPiece chessPieceToPlace = chessPieceFactory.getChessPiece(chessPieceType, row, column);

				// Если фигуру нельзя безопасно разместить на доске, то пропускаем итерацию
				if (!chessPieceToPlace.isSafe(chessBoard)) {
					continue;
				}

				// Добавляем фигуру на доску
				chessBoard.addChessPiece(chessPieceToPlace);
				lastPlacedChessPieces.put(chessPieceToPlace.toString(), chessPieceToPlace);

				if (remainingChessPieces.isEmpty()) {
					// Если фигур больше не осталось, то решение найдено
					numberOfUniqueSolutions++;
				} else {
					// Продолжение поиска решения
					solve();
				}

				// Убираем последнюю фигуру
				chessBoard.removePiece(chessPieceToPlace);
			}

			c = 0;
		}

		// Возвращаем фигуру
		remainingChessPieces.add(0, chessPieceType);
		lastPlacedChessPieces.put(chessPieceType, null);
	}
}

File: src/main/kotlin/ru/timakden/chesspuzzle/ChessPuzzle.kt
package ru.timakden.chesspuzzle

data class ChessPuzzle(
    val rows: Int,
    val columns: Int,
    val kings: Int,
    val queens: Int,
    val rooks: Int,
    val bishops: Int,
    val knights: Int
) {
    var numberOfUniqueSolutions = 0
        private set

    private val board: ChessBoard

    private val lastPlacedPieces = mutableMapOf<String, ChessPiece?>()

    init {
        val remainingChessPieces = mutableListOf<String>()

        repeat(kings) { remainingChessPieces += KING }
        repeat(queens) { remainingChessPieces += QUEEN }
        repeat(rooks) { remainingChessPieces += ROOK }
        repeat(bishops) { remainingChessPieces += BISHOP }
        repeat(knights) { remainingChessPieces += KNIGHT }

        board = ChessBoard(rows, columns, ArrayDeque(remainingChessPieces))
    }

    fun solve() {
        // Take first available piece
        val pieceType = board.remainingPieces.removeFirst()

        var r = 0
        var c = 0

        // Get last placed piece of that type
        lastPlacedPieces[pieceType]?.let {
            // Skip duplicated solutions
            r = it.row
            c = it.column + 1
        }

        for (row in r until board.rows) {
            for (column in c until board.columns) {
                // If cell is occupied then skip current iteration
                if (!board.isCellFree(row, column)) continue

                val pieceToPlace = ChessPieceFactory.getPiece(pieceType, row, column)

                // If this piece can't be safely placed then skip current iteration
                if (!pieceToPlace.isSafe(board)) continue

                // Add piece to board
                board.placedPieces.add(0, pieceToPlace)
                lastPlacedPieces[pieceToPlace.toString()] = pieceToPlace

                if (board.remainingPieces.isEmpty()) numberOfUniqueSolutions++ else solve()

                // Remove last placed piece
                board.placedPieces.remove(pieceToPlace)
            }

            c = 0
        }

        // Return piece to board
        board.remainingPieces.addFirst(pieceType)
        lastPlacedPieces[pieceType] = null
    }
}
